aws/resource_aws_eip.go

```diff
@@ -132,16 +153,22 @@ func resourceAwsEipRead(d *schema.ResourceData, meta interface{}) error {
 		"[DEBUG] EIP describe configuration: %s (domain: %s)",
 		req, domain)
 
-	describeAddresses, err := ec2conn.DescribeAddresses(req)
-	if err != nil {
-		if ec2err, ok := err.(awserr.Error); ok && (ec2err.Code() == "InvalidAllocationID.NotFound" || ec2err.Code() == "InvalidAddress.NotFound") {
-			d.SetId("")
-			return nil
-		}
+	stateConf := &resource.StateChangeConf{
+		Pending:        []string{"notfound"},
+		Target:         []string{"found"},
+		Timeout:        15 * time.Minute,
+		Delay:          10 * time.Second,
+		Refresh:        describeAddressesFunc(ec2conn, req),
+		NotFoundChecks: 90,
+	}
```

Comment:
Since we're retrying on errors, not checking an actual state returned by the API I think `resource.Retry()` would suffice. It has much simpler interface than `StateChangeConf` here and we don't have to make up state names like `notfound` and `found`. 😉 

While 15mins seems fairly high, I'd be ok with such timeout if we only waited in the context of creation, not in `Read`. Only right after creation we can be almost sure the resource _should_ exist.

That `d.SetId("")` is there for a good reason.
It's a promise of Terraform to recover (ideally as quickly as possible) from situation where the user or another tool steps in and deletes resource Terraform has created before.

You can wrap the waiter in `d.IsNewResource()` to satisfy both use cases. 🙂